src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,261 @@
+
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "kafka/server/handlers/alter_partition_reassignments.h"
+
+#include "cluster/errc.h"
+#include "cluster/metadata_cache.h"
+#include "cluster/topics_frontend.h"
+#include "config/node_config.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/alter_partition_reassignments_request.h"
+#include "kafka/protocol/schemata/alter_partition_reassignments_response.h"
+#include "kafka/server/errors.h"
+#include "kafka/server/fwd.h"
+#include "kafka/server/protocol_utils.h"
+#include "model/namespace.h"
+#include "model/timeout_clock.h"
+
+#include <absl/container/flat_hash_set.h>
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <iterator>
+#include <vector>
+
+namespace kafka {
+
+using partitions_request_iterator
+  = std::vector<reassignable_partition>::iterator;
+
+template<typename ResultIter, typename Predicate>
+partitions_request_iterator validate_replicas(
+  partitions_request_iterator begin,
+  partitions_request_iterator end,
+  ResultIter out_it,
+  error_code ec,
+  const ss::sstring& error_message,
+  Predicate&& p) {
+    auto valid_range_end = std::partition(
+      begin, end, std::forward<Predicate>(p));
+
+    if (valid_range_end != end) {
+        vlog(klog.debug, "Invalid replicas: ec {} - {}", ec, error_message);
+    }
+
+    std::transform(
+      valid_range_end,
+      end,
+      out_it,
+      [ec, &error_message](const reassignable_partition& partition) {
+          return reassignable_partition_response{
+            .partition_index = partition.partition_index,
+            .error_code = ec,
+            .error_message = error_message,
+          };
+      });
+    return valid_range_end;
+}
+
+template<typename ResultIter>
+partitions_request_iterator validate_partitions(
```

Comment:
You could write some C++ unit tests for these error conditions. All you would have to do is setup a test with a cluster test fixture and use the raw kafka transport to send requests with bad inputs.

Reply:
I definitely need to add unit tests. My goal is to add those next week so I can focus on getting the feature in this week. If I have time today I'll add some in.